core/src/ballot_box.rs

```diff
@@ -0,0 +1,378 @@
+use spl_math::precise_number::PreciseNumber;
+
+use crate::{constants::PRECISE_CONSENSUS, discriminators::Discriminators, error::TipRouterError};
+
+#[derive(Debug, Clone, PartialEq, Eq, Copy, Zeroable, ShankType, Pod, ShankType)]
+#[repr(C)]
+pub struct Ballot {
+    merkle_root: [u8; 32],
+    max_total_claim: PodU64,
+    max_node_count: PodU64,
+    reserved: [u8; 64],
+}
+
+impl Default for Ballot {
+    fn default() -> Self {
+        Self {
+            merkle_root: [0; 32],
+            max_total_claim: PodU64::from(0),
+            max_node_count: PodU64::from(0),
+            reserved: [0; 64],
+        }
+    }
+}
+
+impl Ballot {
+    pub fn new(root: [u8; 32], max_total_claim: u64, max_num_nodes: u64) -> Self {
+        Self {
+            merkle_root: root,
+            max_total_claim: PodU64::from(max_total_claim),
+            max_node_count: PodU64::from(max_num_nodes),
+            reserved: [0; 64],
+        }
+    }
+
+    pub const fn root(&self) -> [u8; 32] {
+        self.merkle_root
+    }
+
+    pub fn max_total_claim(&self) -> u64 {
+        self.max_total_claim.into()
+    }
+
+    pub fn max_node_count(&self) -> u64 {
+        self.max_node_count.into()
+    }
+}
+
+#[derive(Debug, Clone, Copy, Zeroable, ShankType, Pod, ShankType)]
+#[repr(C)]
+pub struct BallotTally {
+    ballot: Ballot,
+    stake_weight: PodU128,
+    tally: PodU64,
+    reserved: [u8; 64],
+}
+
+impl Default for BallotTally {
+    fn default() -> Self {
+        Self {
+            ballot: Ballot::default(),
+            stake_weight: PodU128::from(0),
+            tally: PodU64::from(0),
+            reserved: [0; 64],
+        }
+    }
+}
+
+impl BallotTally {
+    pub fn new(ballot: Ballot, stake_weight: u128) -> Self {
+        Self {
+            ballot,
+            stake_weight: PodU128::from(stake_weight),
+            tally: PodU64::from(1),
+            reserved: [0; 64],
+        }
+    }
+
+    pub const fn ballot(&self) -> Ballot {
+        self.ballot
+    }
+
+    pub fn stake_weight(&self) -> u128 {
+        self.stake_weight.into()
+    }
+
+    pub fn tally(&self) -> u64 {
+        self.tally.into()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.stake_weight() == 0
+    }
+
+    pub fn increment_tally(&mut self, stake_weight: u128) -> Result<(), TipRouterError> {
+        self.stake_weight = PodU128::from(
+            self.stake_weight()
+                .checked_add(stake_weight)
+                .ok_or(TipRouterError::ArithmeticOverflow)?,
+        );
+        self.tally = PodU64::from(
+            self.tally()
+                .checked_add(1)
+                .ok_or(TipRouterError::ArithmeticOverflow)?,
+        );
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Copy, Zeroable, ShankType, Pod, ShankType)]
+#[repr(C)]
+pub struct OperatorVote {
+    operator: Pubkey,
+    slot_voted: PodU64,
+    stake_weight: PodU128,
+    ballot: Ballot,
```

Comment:
Could be worth just doing a ballot_index here to save the space, where ballot_index points to the location in ballot_tallies that this unique ballot lives - would make this account nearly half the size, and cheaper to do the comparison when looking for the winning operators

Reply:
Deal, nice thought

Reply:
Done